x/distrib/keeper/delegation.go

```diff
@@ -0,0 +1,110 @@
+	// note: necessary to truncate so we don't allow withdrawing more rewards than owed
+	stake := validator.TokensFromSharesTruncated(delegation.GetShares())
+	k.distrKeeper.SetDelegatorStartingInfo(ctx, val, del, types.NewDelegatorStartingInfo(previousPeriod, stake, uint64(ctx.BlockHeight())))
+}
+
+func (k Keeper) withdrawAllDelegationRewards(ctx sdk.Context, delAddr sdk.AccAddress) (sdk.Coins, error) {
+	rewardsTotal := sdk.Coins{}
+	k.stakingKeeper.IterateDelegations(ctx, delAddr, func(_ int64, del stakingtypes.DelegationI) (stop bool) {
+		valAddr := del.GetValidatorAddr()
+		val := k.stakingKeeper.Validator(ctx, valAddr)
+		// end current period and calculate rewards
+		endingPeriod := k.distrKeeper.IncrementValidatorPeriod(ctx, val)
+		rewardsRaw := k.distrKeeper.CalculateDelegationRewards(ctx, val, del, endingPeriod)
+		outstanding := k.distrKeeper.GetValidatorOutstandingRewardsCoins(ctx, del.GetValidatorAddr())
+		// defensive edge case may happen on the very final digits
+		// of the decCoins due to operation order of the distribution mechanism.
+		rewards := rewardsRaw.Intersect(outstanding)
+		if !rewards.IsEqual(rewardsRaw) {
+			logger := k.Logger(ctx)
+			logger.Info(
+				"rounding error withdrawing rewards from validator",
+				"delegator", del.GetDelegatorAddr().String(),
+				"validator", val.GetOperator().String(),
+				"got", rewards.String(),
+				"expected", rewardsRaw.String(),
+			)
+		}
+		// truncate reward dec coins, return remainder to community pool
+		finalRewards, remainder := rewards.TruncateDecimal()
+		k.distrKeeper.SetValidatorOutstandingRewards(ctx, del.GetValidatorAddr(), types.ValidatorOutstandingRewards{Rewards: outstanding.Sub(rewards)})
+		feePool := k.distrKeeper.GetFeePool(ctx)
+		feePool.CommunityPool = feePool.CommunityPool.Add(remainder...)
+		k.distrKeeper.SetFeePool(ctx, feePool)
+		// decrement reference count of starting period
+		startingInfo := k.distrKeeper.GetDelegatorStartingInfo(ctx, del.GetValidatorAddr(), del.GetDelegatorAddr())
+		startingPeriod := startingInfo.PreviousPeriod
+		k.decrementReferenceCount(ctx, del.GetValidatorAddr(), startingPeriod)
+		k.initializeDelegation(ctx, valAddr, delAddr)
+		rewardsTotal = rewardsTotal.Add(finalRewards...)
+		ctx.EventManager().EmitEvent(
+			sdk.NewEvent(
+				types.EventTypeWithdrawRewards,
+				sdk.NewAttribute(sdk.AttributeKeyAmount, finalRewards.String()),
+				sdk.NewAttribute(types.AttributeKeyValidator, val.GetOperator().String()),
+			),
+		)
+		return false
+	})
+	// add coins to user account
+	if !rewardsTotal.IsZero() {
+		withdrawAddr := k.distrKeeper.GetDelegatorWithdrawAddr(ctx, delAddr)
+		err := k.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, withdrawAddr, rewardsTotal)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		baseDenom, _ := sdk.GetBaseDenom()
+		rewardsTotal = sdk.Coins{sdk.Coin{
+			Denom:  baseDenom,
+			Amount: math.ZeroInt(),
+		}}
```

Comment:
I'm not sure honestly, I have this here because it's in the regular withdraw code. I assumed the zero value for Coins would serialize to something unexpected or harder to parse.